File: simulation++/tests/scenario3.py
# Twin scenario 3 (Python side) — midnight-crossing weekly shifts + touching-
# interval merging. Must match scenario3.cpp. Requires the merge changes:
# merge_touching_sorted_intervals applied in HasShifts/IntervalWaiter and
# generate_weekly_shifts taking minutes-of-day pairs.
import sys
from datetime import date, datetime

import salabim as sim
from simulation import env
from simulation.interval import Interval
from simulation.shift_manager import ShiftManager
from simulation.piece import Model, PieceGenerator
from simulation.outlet import Buffer, BufferType
from simulation.sampler import Distribution
from simulation.operator import OperatorGroup, Alternative
from simulation.protocols import (AbortPendingCarriers, NotConstrainedByShift, Conscious)
from simulation.task import Protocols, Scope
from simulation.piece_task import (PieceTaskConfig, ModelConfig, PieceTask, PieceCollectorType)
from simulation.interrupters import NonFlexibleShutdowns
from simulation.judgement_day import ByTime, SimulationStopper

env.trace(True)

model_a = Model("A")

# Mon..Fri: early piece 00:00-06:00 + night piece 22:00-24:00 (in minutes).
# Consecutive working days merge at midnight into 22:00 -> 06:00 night shifts;
# Wednesday is a day off, so Tuesday's night piece stays a 22:00-24:00 stub.
night = [(0.0, 360.0), (1320.0, 1440.0)]
gen_shifts = ShiftManager.generate_weekly_shifts(
    sim_start=datetime(2026, 1, 5),                # a Monday, 00:00
    shifts_per_day=[night] * 5 + [[]] * 2,
    working_days=[True] * 5 + [False] * 2,
    days_off={date(2026, 1, 7)},                   # Wednesday off
    start=date(2026, 1, 5),
    end=date(2026, 1, 11),
)
print("gen_shifts:", " ".join(f"({s.start:g}, {s.end:g})" for s in gen_shifts), file=sys.stderr)

b0 = Buffer("B0", valid_models=[model_a], buffer_type=BufferType.PASSAGE)
gen = PieceGenerator(models_goals={model_a: 40}, shifts=gen_shifts, outlets=[b0])

# an operator whose schedule is two back-to-back shift lists: merged on arrival
og1 = OperatorGroup("og1", capacity=1,
                    shifts=[Interval(360, 840)] + [Interval(840, 1320)],
                    productivity=Distribution(sim.Uniform, 0.9, 1.1))

exit_buffer = Buffer("EXIT", valid_models=[model_a], buffer_type=BufferType.EXIT)

protocols = Protocols(
    pending_carriers_pre_flexible_shutdowns=AbortPendingCarriers(),
    pending_carrier_pre_task_shift_end=AbortPendingCarriers(),
    operator_shift_constraint=NotConstrainedByShift(),
    task_shift_constraint=NotConstrainedByShift(),
    operators_self_conscious=Conscious(),
)
t1_config = PieceTaskConfig(
    task_shifts=[Interval(0, 9000)],
    startup_duration=Distribution(sim.Constant, 3),
    loading_duration=Distribution(sim.Constant, 1),
    startup_operators=Alternative(),
    loading_operators=Alternative(),
    operators=Alternative([(og1, 1)]),
    operator_scope=Scope.PER_BATCH,
    resource_scope=Scope.PER_BATCH,
    min_carriers=1,
    max_capacity=4,
    contiguous_carriers=False,
    independent_carriers=False,
    timeout=400,
    priority=5,
    protocols=protocols,
    models_configs={
        model_a: ModelConfig(duration=Distribution(sim.Uniform, 5, 8),
                             resources=[], min_carrier_capacity=1, max_carrier_capacity=2),
    },
    piece_collector_type=PieceCollectorType.NON_DISCRIMINATING_GREEDY,
)
t1 = PieceTask(config=t1_config, inlets=[b0], outlets=[exit_buffer])
# two shutdown windows touching at 560: must behave as one 500-620 window
NonFlexibleShutdowns(task=t1, intervals=[Interval(500, 560), Interval(560, 620)])

criterion = ByTime(time=9000)
SimulationStopper(criterion=criterion)

env.run(till=100000)

print("=== FINAL STATE ===", file=sys.stderr)
print(f"now={env.now():.6f}", file=sys.stderr)
print("gen_shifts_merged=[" + ", ".join(f"({s.start:g}, {s.end:g})" for s in gen.shifts) + "]", file=sys.stderr)
print("og1_shifts_merged=[" + ", ".join(f"({s.start:g}, {s.end:g})" for s in og1.shifts) + "]", file=sys.stderr)
print("shutdowns_merged=[" + ", ".join(f"({i.start:g}, {i.end:g})" for i in t1.non_flexible_shutdowns.intervals) + "]", file=sys.stderr)
print(f"generated=[{gen.generated[0]}]", file=sys.stderr)
for name, buf in (("B0", b0), ("EXIT", exit_buffer)):
    contents = [(p.id, p.model.name) for p in buf]
    print(f"{name} len={len(buf)} {contents}", file=sys.stderr)
import random
import numpy as np
s = [random.random() for _ in range(3)]
print(f"salabim_stream_next=[{s[0]:.12f}, {s[1]:.12f}, {s[2]:.12f}]", file=sys.stderr)
n = [float(np.random.random_sample()) for _ in range(3)]
print(f"np_stream_next=[{n[0]:.12f}, {n[1]:.12f}, {n[2]:.12f}]", file=sys.stderr)

File: simulation++/tests/scenario3.cpp
// Twin scenario 3 (C++ side) — midnight-crossing weekly shifts + touching-
// interval merging. Must match scenario3.py.
#include "simulation.hpp"

#include <chrono>
#include <cstdio>

using namespace simulation;

int main() {
    auto& e = init(0, false);
    e.trace(true);

    auto* model_a = new Model("A");

    using days_t = ShiftManager::days_t;
    auto day = [](int y, unsigned m, unsigned d) {
        return days_t(std::chrono::year(y) / m / d);
    };
    ShiftManager::DateTime sim_start{day(2026, 1, 5), 0, 0};  // a Monday, 00:00

    std::vector<std::pair<double, double>> night{{0.0, 360.0}, {1320.0, 1440.0}};
    std::vector<std::vector<std::pair<double, double>>> shifts_per_day{
        night, night, night, night, night, {}, {}};
    std::set<long long> days_off{day(2026, 1, 7).time_since_epoch().count()};  // Wednesday off
    Intervals gen_shifts = ShiftManager::generate_weekly_shifts(
        sim_start, shifts_per_day, {true, true, true, true, true, false, false},
        days_off, day(2026, 1, 5), day(2026, 1, 11));
    std::fprintf(stderr, "gen_shifts:");
    for (const auto& s : gen_shifts) std::fprintf(stderr, " (%g, %g)", s->start, s->end);
    std::fprintf(stderr, "\n");

    auto* b0 = new Buffer("B0", {model_a}, BufferType::PASSAGE);
    auto* gen = sim::make<PieceGenerator>({}, std::vector<std::pair<Model*, int>>{{model_a, 40}},
                                          gen_shifts, std::vector<Outlet*>{b0});

    auto* og1 = new OperatorGroup("og1", 1, Intervals{interval(360, 840), interval(840, 1320)},
                                  distribution(DistType::Uniform, {0.9, 1.1}));

    auto* exit_buffer = new Buffer("EXIT", {model_a}, BufferType::EXIT);

    Protocols protocols{
        .pending_carriers_pre_flexible_shutdowns = std::make_shared<AbortPendingCarriers>(),
        .pending_carrier_pre_task_shift_end = std::make_shared<AbortPendingCarriers>(),
        .operator_shift_constraint = std::make_shared<NotConstrainedByShift>(),
        .task_shift_constraint = std::make_shared<NotConstrainedByShift>(),
        .operators_self_conscious = std::make_shared<Conscious>(),
    };
    auto t1_config = std::make_shared<PieceTaskConfig>();
    t1_config->task_shifts = {interval(0, 9000)};
    t1_config->startup_duration = distribution(DistType::Constant, {3});
    t1_config->loading_duration = distribution(DistType::Constant, {1});
    t1_config->startup_operators = Alternative();
    t1_config->loading_operators = Alternative();
    t1_config->operators = Alternative({{{og1, 1}}});
    t1_config->operator_scope = Scope::PER_BATCH;
    t1_config->resource_scope = Scope::PER_BATCH;
    t1_config->min_carriers = 1;
    t1_config->max_capacity = 4;
    t1_config->contiguous_carriers = false;
    t1_config->independent_carriers = false;
    t1_config->timeout = 400;
    t1_config->priority = 5;
    t1_config->protocols = protocols;
    t1_config->models_configs = {
        {model_a, ModelConfig{distribution(DistType::Uniform, {5, 8}), {}, 1, 2}},
    };
    t1_config->piece_collector_type = PieceCollectorType::NON_DISCRIMINATING_GREEDY;
    auto* t1 = sim::make<PieceTask>({}, t1_config, std::vector<Buffer*>{b0},
                                    std::vector<Outlet*>{exit_buffer});
    sim::make<NonFlexibleShutdowns>({}, t1, Intervals{interval(500, 560), interval(560, 620)});

    auto* criterion = sim::make<ByTime>({}, 9000.0);
    sim::make<SimulationStopper>({}, criterion);

    e.run(sim::RunOpts{.till = 100000});

    std::fprintf(stderr, "=== FINAL STATE ===\n");
    std::fprintf(stderr, "now=%.6f\n", e.now());
    std::fprintf(stderr, "gen_shifts_merged=[");
    for (size_t i = 0; i < gen->shifts.size(); ++i)
        std::fprintf(stderr, "%s(%g, %g)", i ? ", " : "", gen->shifts[i]->start, gen->shifts[i]->end);
    std::fprintf(stderr, "]\n");
    std::fprintf(stderr, "og1_shifts_merged=[");
    for (size_t i = 0; i < og1->shifts.size(); ++i)
        std::fprintf(stderr, "%s(%g, %g)", i ? ", " : "", og1->shifts[i]->start, og1->shifts[i]->end);
    std::fprintf(stderr, "]\n");
    std::fprintf(stderr, "shutdowns_merged=[");
    const auto& sdiv = t1->non_flexible_shutdowns->intervals;
    for (size_t i = 0; i < sdiv.size(); ++i)
        std::fprintf(stderr, "%s(%g, %g)", i ? ", " : "", sdiv[i]->start, sdiv[i]->end);
    std::fprintf(stderr, "]\n");
    std::fprintf(stderr, "generated=[%d]\n", gen->generated[0]);
    struct NB { const char* n; Buffer* b; } bufs[] = {{"B0", b0}, {"EXIT", exit_buffer}};
    for (auto [n, b] : bufs) {
        std::fprintf(stderr, "%s len=%zu [", n, (size_t)b->size());
        bool first = true;
        for (auto* c : *static_cast<sim::Queue*>(b)) {
            auto* p = dynamic_cast<Piece*>(c);
            std::fprintf(stderr, "%s('%s', '%s')", first ? "" : ", ", p->id.c_str(), p->model->name.c_str());
            first = false;
        }
        std::fprintf(stderr, "]\n");
    }
    double s1 = sim::random_stream().random(), s2 = sim::random_stream().random(),
           s3 = sim::random_stream().random();
    std::fprintf(stderr, "salabim_stream_next=[%.12f, %.12f, %.12f]\n", s1, s2, s3);
    double n1 = np_random.random_sample(), n2 = np_random.random_sample(),
           n3 = np_random.random_sample();
    std::fprintf(stderr, "np_stream_next=[%.12f, %.12f, %.12f]\n", n1, n2, n3);
    return 0;
}
